src/DecryptionFailureTracker.js

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2018 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+class DecryptionFailure {
+    constructor(failedEventId) {
+        this.failedEventId = failedEventId;
+        this.ts = Date.now();
+    }
+}
+
+export default class DecryptionFailureTracker {
+    // Array of items of type DecryptionFailure. Every `CHECK_INTERVAL_MS`, this list
+    // is checked for failures that happened > `GRACE_PERIOD_MS` ago. Those that did
+    // are added to `failuresToTrack`.
+    failures = [];
+
+    // Every TRACK_INTERVAL_MS (so as to spread the number of hits done on Analytics),
+    // one DecryptionFailure of this FIFO is removed and tracked.
+    failuresToTrack = [];
+
+    // Spread the load on `Analytics` by sending at most 1 event per
+    // `TRACK_INTERVAL_MS`.
+    static TRACK_INTERVAL_MS = 1000;
+
+    // Call `checkFailures` every `CHECK_INTERVAL_MS`.
+    static CHECK_INTERVAL_MS = 5000;
+
+    // Give events a chance to be decrypted by waiting `GRACE_PERIOD_MS` before moving
+    // the failure to `failuresToTrack`.
+    static GRACE_PERIOD_MS = 5000;
+
+    constructor(fn) {
+        if (!fn || typeof fn !== 'function') {
+            throw new Error('DecryptionFailureTracker requires tracking function');
+        }
+
+        this.trackDecryptionFailure = fn;
+    }
+
+    eventDecrypted(e) {
+        if (e.isDecryptionFailure()) {
+            this.addDecryptionFailureForEvent(e);
+        } else {
+            // Could be an event in the failures, remove it
+            this.removeDecryptionFailuresForEvent(e);
+        }
+    }
+
+    addDecryptionFailureForEvent(e) {
+        this.failures.push(new DecryptionFailure(e.getId()));
+    }
+
+    removeDecryptionFailuresForEvent(e) {
+        this.failures = this.failures.filter((f) => f.failedEventId !== e.getId());
+    }
+
+    /**
+     * Start checking for and tracking failures.
+     * @return {function} a function that clears state and causes DFT to stop checking for
+     * and tracking failures.
+     */
+    start() {
+        const checkInterval = setInterval(
+            () => this.checkFailures(Date.now()),
+            DecryptionFailureTracker.CHECK_INTERVAL_MS,
+        );
+
+        const trackInterval = setInterval(
+            () => this.trackFailure(),
+            DecryptionFailureTracker.TRACK_INTERVAL_MS,
+        );
+
+        return () => {
```

Comment:
Any reason? Most libraries I've worked with use this pattern and it has the advantage of not needing to store `check,trackInterval` on the instance.

Reply:
Hmm - just seems more symmetrical to me that of you have a class that you instantiate, both the start & stop methods would be on that class.

From a more practical standpoint, I don't know that storing the timer refs in a closure rather than on the instance is an advantage: this means you can't check if an instance has already been started so if start were called multiple times, you'd end up with multiple sets of timers firing functions that operate on the same object.

Reply:
I would argue that it's easier to spot a situation in which you're potentially going to start something twice if it's returning the function that stops it.

The advantage is that less things have to care about the timers, including the class itself. 

> so if start were called multiple times, you'd end up with multiple sets of timers firing functions that operate on the same object.

If we care about this, both ways of doing it would have to implement ways of preventing that. But given this particular bit of code is only used from the body of `MatrixChat`, I think we can make the same assumption that we make elsewhere that we only call `_onWillStartClient` once per login.

Reply:
I'm not sure that tighter variable scoping within the class should be a justification for making the API of a class a different shape. 

Mostly though, I think this is a matter of consistency. Notifier, UserActivity, Presence, DMRoomMap, ScalarMessaging, ToWidgetPostMessageApi, WidgetMessaging and MatrixActionCreators all use start/stop on the class (or instance). I think the only thing we have where the teardown function is returned by the init function is createDialog, which is because it's tearing down the thing created by that function rather than the object it was called on.

Reply:
>I'm not sure that tighter variable scoping within the class should be a justification for making the API of a class a different shape.

Tighter variable scoping means no looking for whether the timer IDs are being accessed anywhere else. It's an advantage, albeit not a major one. 

>consistency

Okay, fair enough that this is inconsistent with other things with functions called `start` and `stop`. it would be nice to formalise this class of thing at some point though, so that using it is an obvious starting point for things like this.